src/ServiceControl.Audit/Auditing/AuditPersister.cs

```diff
@@ -24,55 +29,157 @@ public void Initialize(IMessageSession messageSession)
             this.messageSession = messageSession;
         }
 
-        public async Task Persist(MessageContext message)
+        public async Task<IReadOnlyList<MessageContext>> Persist(List<MessageContext> contexts)
         {
-            if (!message.Headers.TryGetValue(Headers.MessageId, out var messageId))
+            var stopwatch = Stopwatch.StartNew();
+
+            if (Logger.IsDebugEnabled)
             {
-                messageId = DeterministicGuid.MakeId(message.MessageId).ToString();
+                Logger.Debug($"Batch size {contexts.Count}");
             }
 
-            var metadata = new ConcurrentDictionary<string, object>
+            using (var bulkInsert = store.BulkInsert(options: new BulkInsertOptions
```

Comment:
Since bulkInsert.Dispose() effectively completes the bulk insert, should we be concerned that a random exception halfway through would fall out of the using scope and complete that bulk insert?

Reply:
All the operations are idempotent so I see no big deal here. Still, it is a good comment. I checked the `DisposeAsync` and the `Dispose` implementation of the bulk insert operations. If you explicitly call `DisposeAsync` then `Dispose` does nothing. I saw that ConcurrencyViolation and other exceptions are raised either on `Dispose` or `DisposeAsync`. For now, I removed the `using` block and expanded the try-catch just in case eventhough the `ProcessedMessage` method does a general try-catch and early release the transport transaction